### DISCUSSION TITLE
Design of a complementary IIR filter pair

### DISCUSSION THREAD

**Discussion by RobinSchmidt** [RobinSchmidt/RS-MET#330] [Show and tell]
the last couple of days i was doing some interesting research again. my goal is to design a filter pair (lowpass/highpass) with perfect reconstruction properties, which means that if you add the lowpass and highpass signals together, you get your original signal back exactly. this is obviously useful to create transparent frequency splitters. 

apparently "low + high = original" is equivalent to "high = original - low", so this can always be achieved by just producing the highpass signal by subtracting the lowpass signal from the original. alas, in general, the frequency response of the "highpass" will be a total mess if you just carelessly choose any old lowpass filter. it works well in the 1st order case though, but for anything with higher order, the lowpass filter needs some really careful tuning in order to get a proper highpass by subtraction. 

i have now derived mathematical conditions that ensure that the frequency response of "highpass = original - lowpass" is the mirror image of the frequency response of "lowpass".  these conditions give me constraints for how to design the lowpass. these conditions themselves, though, say nothing about how the frequency response of the lowpass will look like - or if it is even lowpass in nature (after all, i could also start with the highpass and obtain the lowpass by subtraction - the result must be the same due to my mirroring/symmetry constraints). ...i'm now exploring the space of possible mirror-image perfect-reconstruction filter pairs. ...plots will follow...

**Comment by elanhickler** [comment]
ok nice. Frequency splitters are good for mixing, mastering, distortion fx, percussion ensembles, creative fx

**Comment by RobinSchmidt** [comment]
i first design a digital prototype halfband filter (i.e. a filter that has it's cutoff at half the nyquist frequency, so it's at a quarter of the sample rate) and then i intend to use the digital frequency warping formulas that i described here (eq2 lowpass-to-lowpass):

http://www.rs-met.com/documents/dsp/TwoInterpretationsOfFrequencyWarpedTransferFunctions.pdf

this is the magnitude response of the first order (1-pole/1-zero, 1p1z) prototype pair - this is what we currently use in the splitter:

![image](https://user-images.githubusercontent.com/1833099/44878019-12842980-aca6-11e8-9c67-df9a9d5eda77.png)

here's a nice 2nd order (2-pole/2-zero) response pair:

![image](https://user-images.githubusercontent.com/1833099/44878195-87576380-aca6-11e8-9236-cf3fb4d1863c.png)

here's something with 2 poles and 3 zeros, i have intentionally made it a bit high-resonance. in the 2p3z case, there's always a resonance bump, although not always that strong:

![image](https://user-images.githubusercontent.com/1833099/44878321-dbfade80-aca6-11e8-8145-b48427003cf4.png)

i don't know, if that's useful. if, then probably only for special effects. and here is a really weird 4-pole,4-zero response:

![image](https://user-images.githubusercontent.com/1833099/44878360-f59c2600-aca6-11e8-8780-444571772ee2.png)

i was getting much nicer response curves with 4 poles already - unfortunately, the filters with the nice responses were unstable :'-( ...i hope that i can find some nice responses that are stable. even though my symmetry constraints already limit the search space, there are still a couple of tweakables. at least, the 2p2z response (2nd plot) looks like it could be useful for a splitter and improve a little (in terms of steepness/separation) on the 1p1z response of the 1st plot

**Comment by RobinSchmidt** [comment]
hmmm...thinking about it again, i begin to wonder if insisting on the highpass being a mirror image of the lowpass is actually a meaningful constraint to apply. sure, it's nice for making reasonably looking plots - but whether or not such a condition is useful from a sonic point of view is actually an entirely different question.

**Comment by RobinSchmidt** [comment]
Ha - I made some nice progress with that. I now have a perfect-reconstruction 3-way splitter. I first split out a bandpass signal, subtract that from the original, giving a notch signal and then split the notch signal further into low and high bands. The bandpass frequency response is just the regular 2nd order bandpass response (think RBJ cookbook or something) and lowpass and highpass responses are punctured with a notch at the bandpass'es center frequency but look otherwise just like regular 1st order low- and highpass resonses. The nice thin is: if I dial the bandwidth of the bandpass to zero, the mid band just disappears and the low and high band reduce exactly to what would be obtained by the 2-way splitter - the notches kinda become infinitely thin, i.e. disappear.

![image](https://user-images.githubusercontent.com/1833099/178902558-e49274b9-bcc6-48a7-8973-8911802e6d93.png)
(there's some bug in the plotting code leading to a wrong scaling of the freq-axis and offsetting the whole plots)

**Comment by RobinSchmidt** [comment]
For some reason, github seems to have swallowed my last comment, so here it is again. I have a new design for the notchy 3-way splitter which features flatter passbands in all 3 bands:

![image](https://user-images.githubusercontent.com/1833099/179922846-1c70bee4-b76f-411b-807d-f08d03aeb2a1.png)

It's based on the 2-pole/2-zero prototype design (2nd plot in the thread). The steepness is still only 6 dB/oct though, despite being a 2-pole. The additional poles and zeros are used to flatten the passband. From a technical perspective, the lack of steepness seems undesirable, but maybe such shallow splitters can be useful for "musical" multiband effects (compressor, distortion, etc.).

**Comment by elanhickler** [comment]
yea, I wonder if a shallow splitter will create a more cohesive multiband
effect. Sometimes multiband effects' filters are steep in that you can hear
the individual harmonic sections due to the effects doing something very
different per split. So if it's shallow, maybe that will be more mixed and
cohesive... if I understand correctly.

On Wed, Jul 20, 2022 at 12:32 AM Robin Schmidt ***@***.***>
wrote:

> For some reason, github seems to have swallowed my last comment, so here
> again. I have a new design of the notchy 3-way splitter which features
> flatter passbands:
>
> [image: image]
> <https://user-images.githubusercontent.com/1833099/179922846-1c70bee4-b76f-411b-807d-f08d03aeb2a1.png>
>
> It's a 2-pole/2 zero design - the steepness is still only 6 dB/oct though.
> The additional poles and zeros are used to flatten the passband. From a
> technical perspective, the lack of steepness seem undesirable, but maybe
> such shallow splitters can be useful for "musical" multiband effects
> (compressor, distortion, etc.).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/RobinSchmidt/RS-MET/discussions/330#discussioncomment-3186718>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACFRXH4RASECXCVPUCME6Y3VU6TO5ANCNFSM5JLADZRA>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
-Elan
www.elanhickler.com
www.soundemote.com

**Comment by RobinSchmidt** [comment]
Yes, exactly that was my idea as well. Although I tend to think that 6 dB/oct might be a bit too shallow. Maybe something between 12 and 24 dB would be nice to have (optionally, of course). Of course, this can be done with Linkwitz/Riley filters which I already have in the library for a decade - but these do not have the perfect reconstruction property: if you split, do nothing, then recombine, you'll get an allpass filter, i.e. a flat magnitude response but messed up ("dispersed") phases. What I want here is nice responses (with good overall shape and LP/HP being mirror images of each other), perfect reconstruction *and* I want to do it with IIR filters...I'm not sure if that's even possible. This kind of stuff is typically done with FIR filters, I think. But I keep trying. At least, I have now a somewhat improved 6 dB/oct perfect-reconstruction splitter...and it may have 3 ways....although the notches in the lo/hi bands may look strange....but the more I think about it, the more sense they seem to make. For a sinusoid exactly at the bandpass center, *all* of the signal will go into in the mid channel and nothing will leak into the lo/hi channels. That kinda implies the notch. Of course one might also argue that one prefers a bit of leakage for the sake of a more regular and smoother freq-response for the high/lowpass filters.

Oh and of course, when we split the resulting bands further using the same technique, we can make splitters with an arbitrary number of bands (as I did back then in the multiband compressor)
